server/cmd_selected.go

```diff
@@ -100,10 +100,14 @@ func (cmd *Expunge) Handle(conn Conn) error {
 			done <- conn.WriteResp(res)
 		})()
 
-		// Iterate sequence numbers from the last one to the first one, as deleting
-		// messages changes their respective numbers
-		for i := len(seqnums) - 1; i >= 0; i-- {
-			ch <- seqnums[i]
+		// Send sequence numbers to channel, and check if conn.WriteResp() finished early.
+		for _, seqnum := range seqnums {
```

Comment:
I reverted back to sending them in reverse.  I do not think it matters.  The backend can do the deletes in any order as long as the sequence numbers match the removed messages.

Reply:
One weird go-imap implementation detail is that a single expunge response object can generate multiple EXPUNGE updates and each EXPUNGE update changes sequence numbers of all messages after deleted one.

Consider that example with three messages deleted (A, B, C with seqnums 1, 2, 3):

```
* EXPUNGE 1
```
This tells client that first message (A) in the mailbox is gone. After that, message B seqnum is 1 and message C seqnum is 2.

```
* EXPUNGE 2
```
This tells client that second message in the mailbox is gone. According to the previous statement, that is message C, apparently. So we now have only message B left in mailbox with seqnum 1.

```
* EXPUNGE 3
```
This tells client that third message in the mailbox is gone. This is neither message B nor any other message we had before. Perhaps it will be completely unrelated message or, perhaps, it will crash the client due to out-of-range array access if there was only 3 messages in the mailbox before. 

RFC 3501 mentions this: https://tools.ietf.org/html/rfc3501#section-7.4.1.